### PR TITLE
22194/Implemented script for docker

### PR DESCRIPTION
**Pull request by Chepa14**

### CODE REVIEW

File: components/data_tunnel/data_tunnel.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        data={
+            "user": user_id,
+            "notebook": component_id,
+            "polygon": area,
+            "date_from": start_date,
+            "date_to": end_date
+        },
+        params=auth_param
+    )
+    print(request)
```
Comment (adolgarev) [inline components/data_tunnel/data_tunnel.py:136]: Use python logging

**Review by adolgarev — COMMENTED** [review]
You should use https://openapi-generator.tech/docs/generators/python/ that is also based on urllib3. Thus you can use https://urllib3.readthedocs.io/en/stable/reference/urllib3.util.html#urllib3.util.Retry

Overall use OOP

File: components/data_tunnel/data_tunnel.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+@retry_request()
+def wait_for_request_success_finish(url, params) -> (bool, str):
+    response = requests.get(url, params=params, timeout=REQUEST_TIMEOUT)
+    if not response.status_code == 200:
+        raise Exception(f"Check created Request error - status code {response.status_code}")
+    curr_request = response.json()
+    if not curr_request.get("calculated"):
+        if curr_request.get("error") and curr_request.get("finished_at"):
+            return False, curr_request.get("error")
+        raise Exception("Not finished yet!")
```
Comment (adolgarev) [inline components/data_tunnel/data_tunnel.py:65]: This doesn't relate to @retry_request, this is polling

File: components/data_tunnel/data_tunnel.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+def pull_results(created_request_id, url, params):
+    response = requests.get(url, params=params, timeout=REQUEST_TIMEOUT)
+    if not response.status_code == 200:
+        raise Exception(f"Pull results error  - status code {response.status_code}")
+    paths = []
+    for result in response.json():
+        if result.get("request") == created_request_id:
+            paths.append(result.get("filepath"))
+    if not paths:
+        raise Exception("No results yet!")
```
Comment (adolgarev) [inline components/data_tunnel/data_tunnel.py:79]: If request finished and no results they won't appear, no point re-requesting

File: components/data_tunnel/data_tunnel.py (hunk 1 above)
Comment (sviridovartem) [reply components/data_tunnel/data_tunnel.py]: done

File: components/data_tunnel/data_tunnel.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        data={
+            "user": user_id,
+            "notebook": component_id,
+            "polygon": area,
+            "date_from": start_date,
+            "date_to": end_date
+        },
+        params=auth_param
+    )
+    log.info(f"request: {request}")
```
Comment (adolgarev) [inline components/data_tunnel/data_tunnel.py:58]: additional_parameter is missing

File: components/data_tunnel/utils.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
+import json
+import time
+import os
+
+RETRY_INTERVAL = 2880  # num of intervals -> every minute
+RETRY_LIMIT_DAYS = 172800.0  # 2 days in seconds
```
Comment (adolgarev) [inline components/data_tunnel/utils.py:6]: Then it is RETRY_LIMIT_SECONDS

**Review by adolgarev — COMMENTED** [review]
Lets call it remote executor

**Comment by sviridovartem** [comment]
> remote executor

done

File: components/data_tunnel/utils.py (hunk 5 above)
Comment (sviridovartem) [reply components/data_tunnel/utils.py]: done

File: components/remote_executor/geoapp_client.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+            time.sleep(60)
+        return False, f"Not finished during {RETRY_LIMIT_SECONDS} seconds"
+
+    def pull_results(self, created_request_id):
+        url = self.api_endpoint + "api/results"
+        start_time = time.time()
+        paths = []
+        self.log.info(
+            f"pull_results started: {time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(start_time))}")
+        while time.time() < start_time + RETRY_LIMIT_SECONDS:
```
Comment (adolgarev) [inline components/remote_executor/geoapp_client.py:95]: No need for retry here

File: components/remote_executor/geoapp_client.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+    def wait_for_request_success_finish(self, request_id):
+        url = self.api_endpoint + f"api/request/{request_id}"
+        start_time = time.time()
+        self.log.info(
+            f"wait_for_request_success started: {time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(start_time))}")
+        while time.time() < start_time + RETRY_LIMIT_SECONDS:
+            response = self.http.request('GET', url, fields=self.params)
+            curr_request = json.loads(response.data.decode())
+            if not curr_request.get("calculated"):
```
Comment (adolgarev) [inline components/remote_executor/geoapp_client.py:78]: Look for finished_at. If it is set then look for calculated

File: components/remote_executor/remote_executor.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+    data = {
+        "user": user_id,
+        "notebook": component_id,
+        "polygon": area,
+        "date_from": start_date if start_date else '',
+        "date_to": end_date if end_date else ''
+    }
+    if additional_parameter_name:
+        data[additional_parameter_name] = additional_parameter
```
Comment (adolgarev) [inline components/remote_executor/remote_executor.py:29]: `data['additional_parameter'] = additional_parameter_value`

File: components/remote_executor/geoapp_client.py (hunk 7 above)
Comment (sviridovartem) [reply components/remote_executor/geoapp_client.py]: done

File: components/remote_executor/remote_executor.py (hunk 8 above)
Comment (sviridovartem) [reply components/remote_executor/remote_executor.py]: done

File: components/remote_executor/geoapp_client.py (hunk 6 above)
Comment (sviridovartem) [reply components/remote_executor/geoapp_client.py]: done

File: components/remote_executor/geoapp_client.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+        return False, f"Not finished during {RETRY_LIMIT_SECONDS} seconds"
+
+    def pull_results(self, created_request_id):
+        url = self.api_endpoint + "api/results"
+        start_time = time.time()
+        paths = []
+        self.log.info(
+            f"pull_results started: {time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(start_time))}")
+        while time.time() < start_time + RETRY_LIMIT_SECONDS:
+            response = self.http.request('GET', url, fields=self.params)
```
[the remark this answers is not in the document]
Comment (sviridovartem) [reply components/remote_executor/geoapp_client.py]: done

**Review by adolgarev — COMMENTED** [review]
You should pass apikey to every request. We are using https://github.com/QuantuMobileSoftware/geoap/blob/master/webapplication/user/authentication.py which allows to pass apikey in an url, this is used e.g. in QGIS. Here it is more comfortable to pass apikey in a header, see https://www.django-rest-framework.org/api-guide/authentication/#tokenauthentication

Also please try to run locally

**Comment by sviridovartem** [comment]
> You should pass apikey to every request. We are using https://github.com/QuantuMobileSoftware/geoap/blob/master/webapplication/user/authentication.py which allows to pass apikey in an url, this is used e.g. in QGIS. Here it is more comfortable to pass apikey in a header, see https://www.django-rest-framework.org/api-guide/authentication/#tokenauthentication
> 
> Also please try to run locally

Moved Encoding and token to headers, removed params
fixed bug with status code
tested locally on TCI_NDVI and ForestAGB -> works

**Pull request merged**